src/main/java/luke/TaskList.java

```diff
@@ -0,0 +1,229 @@
+     */
+    public boolean isOutOfBounds(int index) {
+        return (index <= 0 || index > tasks.size());
+    }
+
+    /**
+     * Checks if tasks is empty.
+     *
+     * @return True if tasks is empty, False if tasks is not empty.
+     */
+    public boolean isEmpty() {
+        return tasks.isEmpty();
+    }
+
+    /**
+     * Adds a task to the list of tasks.
+     *
+     * @param taskType The type of the task to be added.
+     * @param taskName The name of the task to be added.
+     * @param taskDate A string containing the dates of the task (E.g deadline or start and end date).
+     * @return True as task have been added without any errors, False if there are errors when adding the task.
+     */
+    public boolean addTask(String taskType, String taskName, String taskDate) {
+        Task newTask;
+        switch (taskType) {
+        case "todo":
+            newTask = new ToDo(taskName, newTaskID);
+            break;
+        case "deadline":
+            if (taskDate == null) {
+                return false;
+            }
+            newTask = new Deadline(taskName, newTaskID, taskDate.trim());
+            break;
+        default:
+            if (taskDate == null) {
+                return false;
+            }
+            String startDate = StringManipulation.getFirstDetail(taskDate);
+            String endDate = StringManipulation.removeFirstDetail(taskDate);
+
+            if (startDate == null || endDate == null) {
+                return false;
+            }
+
+            newTask = new Event(taskName, newTaskID, startDate.trim(), endDate.trim());
+            break;
+        }
+        tasks.put(newTaskID, newTask);
+        newTaskID += 1;
+        updateSerialNumbers();
+        return true;
+    }
+
+    /**
+     * Given the serial number of a task, this function marks the corresponding task as completed.
+     *
+     * @param toMarkSerialNumber The serial number of the task to be marked.
+     */
+    public void markTask(int toMarkSerialNumber) {
+        int toMarkID = serialNumbers.get(toMarkSerialNumber);
+        tasks.get(toMarkID).markTask();
+    }
+
+    /**
+     * Given the serial number of a task, this function unmarks a task.
+     *
+     * @param toUnmarkSerialNumber The serial number of the task to be unmarked.
+     */
+    public void unmarkTask(int toUnmarkSerialNumber) {
+        int toUnmarkID = serialNumbers.get(toUnmarkSerialNumber);
+        tasks.get(toUnmarkID).unmarkTask();
+    }
+
+    /**
+     * Given the ID of a task, this function deletes a task.
+     *
+     * @param toDeleteSerialNumber The ID of the task to be deleted.
+     */
+    public void deleteTask(int toDeleteSerialNumber) {
+        int toDeleteID = serialNumbers.get(toDeleteSerialNumber);
+        tasks.remove(toDeleteID);
+        updateSerialNumbers();
+    }
+
+    public ArrayList<Task> findTask(String toFind) {
+        ArrayList<Task> found = new ArrayList<>();
```

Comment:
Code Standard: Use plural forms for collection of objects.

Can consider renaming to `foundTasks`.